### ISSUE TITLE
Improvements: skeletons

### ISSUE THREAD

**Issue by martinschayna** [recharts/recharts#1707]
Maybe it's out of scope, but it would be nice to have a skeleton of a chart, to be rendered when data are temporarily unknown, probably based on a new property. 

It shouldn't have anything to do with the real chart, it will be such a feeling of it, ie. some vertical bars for a bar chart, lines for a line chart, etc.

**Comment by semoal** [comment]
Correct me If I'm wrong, but with React.lazy(() => import()), and Suspense you can get a nice placeholder while this component is still loading/or just heavier than normal. 

Implement this on the side of Recharts is practically impossible to catch all the situations that could happen.

**Comment by martinschayna** [comment]
Yes, I can use Suspense or less sophisticated loading state handling. What I had on my mind was an implementation of Suspense `fallback` content, instead of "Loading..." text I want to show a Bar chart as a grey skeleton or Line chart skeleton, etc. based on same Rechart components properties excluded `data` property. 

Something like that:
```
{data 
  ? <LineChart ...props data={data}> ... </LineChart>
  : <LineChart ...props skeleton={true}/> ... </LineChart>
}
```

**Comment by xile611** [comment]
@martinschayna Thanks! But we have no plan to add skeletons now.

**Comment by 7hibault** [comment]
This is a feature I would find useful :)

**Comment by trsiddiqui** [comment]
Me too

**Comment by sli** [comment]
Would be especially nice with pie charts, which simply render nothing but a blank space if your data is currently all zeros. Since there's no documentation on how to use Customized, that's not really a solution.

**Comment by slaat** [comment]
Is there any news on this?

**Comment by BjornGjorv** [comment]
Would also love this feature!

**Comment by LouisMuriuki** [comment]
we want this feature

**Comment by based64-eth** [comment]
wen?

**Comment by Danoodli** [comment]
we still want it!

**Comment by ckifer** [comment]
The appetite for this issue confuses me a bit - there are a lot of closed issues in this repo that I agree with reopening, but I don't think this is one of them (at least with its current title). Styled skeletons are out of scope for this library.

Are there any other charting libraries that provide a skeleton loader or a loading state in general?

I would be down to support something like this https://api.highcharts.com/class-reference/Highcharts.Chart#showLoading, but in general something like a skeleton will be very style/UI library specific. [MUI Skeleton](https://mui.com/material-ui/react-skeleton/) for example won't match the style of someone's [react-loading-skeleton](https://www.npmjs.com/package/react-loading-skeleton). 

Conditional rendering by the implementer offers control that you just won't have with an internal recharts implementation.

My preference here would be to support a slot for `loading: Boolean` and `loader: React.ReactElement` without any specific implementation. Maybe "Loading..." text by default.

**Comment by LouisMuriuki** [comment]
@ckifer The idea basically is exactly as you have said it, i personally would like to have a loading indicator instead of a blank section as my chart loads, something that can interact with the chart seemlesly providing a good ui that matches the chart. The only thing that came in mind is a chart skeleton and am pretty sure that is why more people are asking for the same.
But a loading prop for now would be probably alright since there isn't any at the moment

**Comment by ckifer** [comment]
Cool - I'm going to re-open and adjust the title a bit

**Comment by vikram2000b** [comment]
Any update on this issue? Do we have this feature now?

**Comment by ckifer** [comment]
No movement on this no, I'd consider this pretty low priority. I still stand by my original comment - this can be done outside of recharts without much issue

**Comment by andresmoschini** [comment]
It would be nice to have something like this out of the box:

https://github.com/user-attachments/assets/185fef26-2875-478d-97e3-6c689434aa70

**Comment by Fefedu973** [comment]
It would be really nice !

**Comment by julianklumpers** [comment]
With tailwind and tailwind-animate it is very easy. conditionally render a Cell with `className` animate-pulse and fill gray 

```ts
<Bar
  background={{ fill: "#eee", opacity: 0.5 }}
  dataKey="y"
  radius={[5, 5, 0, 0]}
  tooltipType="none"
>
  <LabelList
    fill="#333"
    fontWeight="bold"
    formatter={(v: number) => `${v}%`}
    position="top"
  />
  {loading
    ? chartData.map(d => (
      <Cell className="animate-pulse" fill="gray" key={d.label} />
    ))
    : null}
</Bar>
```

**Comment by kyrylolvov** [comment]
> It would be nice to have something like this out of the box:
> 
>  2024-08-29_15-13-16.mp4

```
<Bar
  fill={chartConfig.apiCalls.color}
  maxBarSize={40}
  dataKey="apiCalls"
  radius={[4, 4, 0, 0]}
  isAnimationActive={false}
  className={cn("fill-primary-foreground", isChartDataEmpty && "fill-border/30", isLoading && "fill-border/75 animate-pulse")}
/>
```
